src/pam_google_authenticator.c

```diff
@@ -1493,6 +1513,125 @@ static int check_timebased_code(pam_handle_t *pamh, const char*secret_filename,
   return 1;
 }
 
+/* 
+ * Add a 'config' variable that says we logged in from a particular place
+ * at a particular time.  Only remembers the last 10 logins, which 
+ * are replaced in LRU order.
+ *
+ * Returns 0 on success.
+ */
+int
+update_logindetails(pam_handle_t *pamh, const Params *params, char **buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  time_t oldest = now;
+  int oldest_index = 0;
+  int i;
+  char *line;
+  unsigned long when;
+  char name[] = "LAST ";
+  char *host;
+
+  if (rhost == NULL) {
+    return -1;
+  }
+
+  for (i = 0; i < 10; i++) {
+    name[4] = i + '0';
+    line = get_cfg_value(pamh, name, *buf);
+    if (line == &oom) {
+      /* Fatal! */
+      return -1;
+    }
+
+    if (!line) {
+      /* Make first empty line the oldest */
+      if (oldest) {
+        oldest_index = i;
+        oldest = 0;
+      }
+      continue;
+    }
+
+    if (sscanf(line, " %m[0-9a-fA-F:.] %lu ", &host, &when) != 2) {
+      log_message(LOG_ERR, pamh, "Malformed LAST%d line\n", i);
+      continue;
+    }
+
+    if (!strcmp(host, rhost)) {
+      break;
+    }
+        
+    if (when < oldest) {
+      oldest_index = i;
+      oldest = when;
+    }
+
+    free(line);
+    free(host);
+  }
+
+  if (i == 10) {
+    name[4] = oldest_index + '0';
+  } else {
+    free(line);
+    free(host);
+  }
+
+  int len = strlen(rhost) + 100;
+  char *value = calloc(len, 1);
+
+  snprintf(value, len - 1, "%s %lu", rhost, (unsigned long)now);
+  set_cfg_value(pamh, name, value, buf);
+
+  free(value);
+
+  return 0;
+}
+
+/*
+ * Return non-zero if the last login from the same host as this one was 
+ * successfully authenticated within the grace period.
+ */
+int
+within_grace_period(pam_handle_t *pamh, const Params *params, char **buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  const time_t grace = params->grace_period;
+  int i;
+  char *line;
+  char name[] = "LAST ";
+  char *p;
```

Comment:
```suggestion
```

Reply:
More complex than that.  Don't really want to reinitialise `name[]` every time through the loop. so it'll end up static.  Early termination means a memory leak unless `line` is available outside to `free()`.
(BTW, I spotted the leak because of your comments).